ateam_kenobi/src/path_planning/path_planner.cpp

```diff
@@ -352,4 +372,50 @@ void PathPlanner::trimPathAfterCollision(
   }
 }
 
+std::optional<ateam_geometry::Point> PathPlanner::findLastCollisionFreePoint(
+  const ateam_geometry::Point & start, const ateam_geometry::Point & goal, const World & world,
+  std::vector<ateam_geometry::AnyShape> & obstacles,
+  const PlannerOptions & options)
+{
+  const auto direction_vector = start - goal;
+  const auto segment_length = std::sqrt(direction_vector.squared_length());
+  const auto step_vector = direction_vector * (options.collision_check_resolution / segment_length);
+  const int step_count = segment_length / options.collision_check_resolution;
+  for (auto step = 0; step < step_count; ++step) {
+    const auto state = goal + (step * step_vector);
+    if (isStateValid(state, world, obstacles, options)) {
+      return state;
+    }
+  }
```

Comment:
Should the direction_vector here be `goal - start` if we start at the goal in the loop starting at 384 and go backwards towards the start?

Reply:
`start - goal` gives the vector from `goal` to `start`.
One way to sanity check this is to look at the end iteration for the loop. `step * step_vector` will equal `start - goal`. Doing the substitution on the initializer for `state`, you get `goal + start - goal`, which reduces down to just `start`, which is where we wanted to end up.
If we flipped it to `goal - start`, that end case would result in `goal + goal - start` or `(2 * goal) - start`.